8316996: Catalog API Enhancement: add a factory method

JoeWang-Java opened the pull request:
Add a new factory method so that a CatalogResolver can be created with a resolve property on top of the Catalog object.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8316997](https://bugs.openjdk.org/browse/JDK-8316997) to be approved

### Issues
 * [JDK-8316996](https://bugs.openjdk.org/browse/JDK-8316996): Catalog API Enhancement: add a factory method (**Enhancement** - P4)
 * [JDK-8316997](https://bugs.openjdk.org/browse/JDK-8316997): Catalog API Enhancement: add a factory method (**CSR**)


### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [89b6b7ea](https://git.openjdk.org/jdk/pull/16045/files/89b6b7ea9b6787f7d2191784491b9c15021531f5)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [89b6b7ea](https://git.openjdk.org/jdk/pull/16045/files/89b6b7ea9b6787f7d2191784491b9c15021531f5)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16045/head:pull/16045` \
`$ git checkout pull/16045`

Update a local copy of the PR: \
`$ git checkout pull/16045` \
`$ git pull https://git.openjdk.org/jdk.git pull/16045/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16045`

View PR using the GUI difftool: \
`$ git pr show -t 16045`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16045.diff">https://git.openjdk.org/jdk/pull/16045.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16045#issuecomment-1747765492)

naotoj commented on src/java.xml/share/classes/javax/xml/catalog/CatalogManager.java:
```diff
@@ -85,6 +88,34 @@ public static CatalogResolver catalogResolver(Catalog catalog) {
+     * and the value of the {@link CatalogFeatures.Feature#RESOLVE RESOLVE} property.
+     *
+     * @param catalog the catalog instance
+     * @param resolve the value of the {@link CatalogFeatures.Feature#RESOLVE RESOLVE}
+     * property that overrides the previous setting used for creating the {@code catalog}
+     * object. The supported values are: {@code strict}, {@code continue},
+     * and {@code ignore}. {@code null} may be specified to indicate that the
+     * {@code catalog} object's current {@link CatalogFeatures.Feature#RESOLVE RESOLVE}
+     * value remains unchanged.
```
Supported values explanation and the behavior on `null` string may be moved into the method description body, as it explains how this method is supposed to work.

naotoj commented on src/java.xml/share/classes/javax/xml/catalog/Util.java:
```diff
@@ -267,4 +269,34 @@ static void validateFeatureInput(Feature f, String value) {
+     * @param base the base, that is the system id of the catalog within the
+     * Catalog implementation
+     * @param uri the specified uri
+     * @return Returns the absolute form of the specified uri
```
"Returns" may be redundant. Anyway, can this `Util` class be `final` and have private constructor not to be instantiated?

naotoj commented on src/java.xml/share/classes/javax/xml/catalog/Util.java:
```diff
@@ -267,4 +269,34 @@ static void validateFeatureInput(Feature f, String value) {
+            }
+            return temp;
+        } catch (MalformedURLException ex) {
+            // no action, shouldn't happen as the base has already been validated
```
If this "shouldn't happen", is that considered an Error, which should be reported?

naotoj commented on src/java.xml/share/classes/javax/xml/catalog/CatalogManager.java:
```diff
@@ -85,6 +88,34 @@ public static CatalogResolver catalogResolver(Catalog catalog) {
+     * @return an instance of a {@code CatalogResolver}
+     * @throws IllegalArgumentException if the value of the {@code resolve} property is
+     * not a supported value or {@code null}.
+     *
```
`@throws NPE` for `catalog==null` is needed

JoeWang-Java commented:
Thanks Naoto. Addressing review comments. As we discussed offline, as we got CSR approval, we'll keep the javadoc in @param as a style choice. Added NPE for catalog; Fixed the redundant "returns"; Added error report. 
Furthermore, after consulting with Daniel, modified the Util method with URL instead of URI/Path as the former is a more straight-forward solution requiring no additional manipulation.

naotoj commented on test/jaxp/javax/xml/jaxp/unittest/catalog/CatalogTest.java:
```diff
@@ -506,7 +506,7 @@ public void testRewriteUri() throws Exception {
     @Test(expectedExceptions = NullPointerException.class)
     public void testFeatureNull() {
-        CatalogResolver resolver = CatalogManager.catalogResolver(null, null);
+        CatalogResolver resolver = CatalogManager.catalogResolver((CatalogFeatures)null, (URI)null);
```
I misunderstood this change, as I thought this was for testing `null` for the *new* method. Should we have the equivalent one for the new method?

JoeWang-Java replied:
True, added a test. Thanks.

naotoj reviewed (approved):
Looks good. Thanks for making the changes

naotoj commented on src/java.xml/share/classes/javax/xml/catalog/CatalogResolver.java:
```diff
@@ -235,4 +235,53 @@ public InputStream resolveEntity(String publicId, String systemId,
+         * @return the action type
+         */
+        static public NotFoundAction getType(String resolve) {
+            for (NotFoundAction type : NotFoundAction.values()) {
+                if (type.toString().equals(resolve)) {
+                    return type;
+                }
+            }
```
Instead of looping through all the elements, `Enum.valueOf()` can be used (and it appropriately throws IAE if not found).

JoeWang-Java replied:
Thanks. Added IAE. Enum.valueOf() is nice. However, in this case, the Enum type is mapped to the resolve property that was defined to be case sensitive. It would therefore require validating the case and then convert it to upper case before calling valueOf.

naotoj replied:
That is correct. Thanks for checking

naotoj reviewed (approved):
LGTM

JoeWang-Java commented:
/integrate
